### PR TITLE
[Optimizer] Remove single identity transpose from the graph

### PR DESCRIPTION
**Pull request by tlepley-cadence**
We've got some ONNX models with an 'identity' transpose. It was not automatically removed by Glow. This PR removes the transpose node in such a case.

### CODE REVIEW

**Comment by tlepley-cadence** [comment]
@jfix71 Sure. Please note that tomorrow is public holidays in France, so I'll work on this Thursday.

File: lib/Optimizer/GraphOptimizer.cpp [hunk 1]
```diff
@@ -108,6 +108,18 @@ static void DCE(Function *F) {
   }
 }
 
+/// \returns true if the \p shuffle corresponds to an identity operation, false
+/// otherwise.
+static bool isIdentityShuffle(llvm::ArrayRef<unsigned> shuffle) {
+  // Check if the combined masks are the identity mask.
```
Comment (rdzhabarov) [inline lib/Optimizer/GraphOptimizer.cpp]: seems like stale comment ;)

Comment (tlepley-cadence) [reply lib/Optimizer/GraphOptimizer.cpp]: oops :) I just removed it.

File: tests/unittests/graphOptzTest.cpp [hunk 2]
```diff
@@ -289,11 +289,25 @@ TEST_F(GraphOptz, cancelTwoTransposes) {
...
+                                VisibilityKind::Public, false);
+  Node *T = F_->createTranspose("transpose", A, {0, 1, 2, 3});
+  Node *K = F_->createRELU("relu", T);
+  F_->createSave("ret", K);
+
+  EXPECT_EQ(F_->getNodes().size(), 3);
+
+  ::glow::optimize(F_, CompilationMode::Infer);
+
+  EXPECT_EQ(F_->getNodes().size(), 2);
```
Comment (qcolombet) [inline tests/unittests/graphOptzTest.cpp]: Could you check the actual shape of the resulting graph?
I.e., K's argument is A.

Comment (rdzhabarov) [reply tests/unittests/graphOptzTest.cpp]: fwiw, we do not follow this good practice in the majority of the existing tests, unfortunately :(

File: tests/unittests/graphOptzTest.cpp [hunk 3]
```diff
@@ -289,11 +289,25 @@ TEST_F(GraphOptz, cancelTwoTransposes) {
   EXPECT_EQ(F_->getNodes().size(), 2);
 }
 
+TEST_F(GraphOptz, sinkIdentityTranspose) {
```
Comment (rdzhabarov) [inline tests/unittests/graphOptzTest.cpp]: nit: there is no actual sink happening here it's just elimination of identity transpose.

**Comment by nadavrot** [comment]
Nice!

File: tests/unittests/graphOptzTest.cpp (hunk 2 above)
Comment (tlepley-cadence) [reply tests/unittests/graphOptzTest.cpp]: Done.

File: tests/unittests/graphOptzTest.cpp (hunk 3 above)
Comment (tlepley-cadence) [reply tests/unittests/graphOptzTest.cpp]: Renamed `remove`. For me `sink` was equivalent to `remove`. @rdzhabarov what is the actual meaning of  `sink` in the context of Glow ?

Comment (rdzhabarov) [reply tests/unittests/graphOptzTest.cpp]: @tlepley-cadence sorry for delay, I'm OOO today.
sink means moving a node below the other node, check `Sink Transpose below batch RELU nodes.` in the graph optimizer for example.

**Comment by nadavrot** [comment]
Congrats!

**Pull request merged**